Mathlib/MeasureTheory/Measure/WithDensityFinite.lean

```diff
@@ -0,0 +1,236 @@
+/-
+Copyright (c) 2024 Rémy Degenne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémy Degenne
+-/
+import Mathlib.MeasureTheory.Decomposition.RadonNikodym
+
+/-!
+# s-finite measures can be written as `withDensity` of a finite measure
+
+If `μ` is an s-finite measure, then there exists a finite measure `μ.toFinite` and a measurable
+function `densityToFinite μ` such that `μ = μ.toFinite.withDensity μ.densityToFinite`.
+
+That measure is not unique, and in particular our implementation leads to `μ.toFinite ≠ μ` even if
+`μ` is finite.
+
+## Main definitions
+
+In all these definitions and the results below, `μ` is an s-finite measure (`SFinite μ`).
+
+* `MeasureTheory.Measure.toFinite`: a finite measure with `μ ≪ μ.toFinite` and `μ.toFinite ≪ μ`.
+* `MeasureTheory.Measure.densityToFinite`: a measurable function such that
+  `μ = μ.toFinite.withDensity μ.densityToFinite`.
+* `MeasureTheory.Measure.toSigmaFinite`: a sigma-finite measure with `μ ≪ μ.toSigmaFinite` and
+  `μ.toSigmaFinite ≪ μ`. If `μ` is sigma-finite, `μ.toSigmaFinite = μ`.
+* `MeasureTheory.Measure.densityToSigmaFinite`: a measurable function such that
+  `μ = μ.toSigmaFinite.withDensity μ.densityToSigmaFinite`. This function takes only two values,
+  `1` and `∞`.
+* `MeasureTheory.Measure.densitySigmaFiniteToFinite`: a measurable function such that
+  `μ.toSigmaFinite = μ.toFinite.withDensity μ.densitySigmaFiniteToFinite`.
+
+## Main statements
+
+* `absolutelyContinuous_toFinite`: `μ ≪ μ.toFinite`.
+* `toFinite_absolutelyContinuous`: `μ.toFinite ≪ μ`.
+* `withDensity_densitytoFinite`: `μ.toFinite.withDensity μ.densityToFinite = μ`.
+* `withDensity_densityToSigmaFinite`: `μ.toSigmaFinite.withDensity μ.densityToSigmaFinite = μ`.
+* `toSigmaFinite_eq_self_iff`: `μ.toSigmaFinite = μ ↔ SigmaFinite μ`.
+
+-/
+
+open scoped ENNReal
+
+namespace MeasureTheory
+
+variable {α : Type*} {mα : MeasurableSpace α}
+
+/-- A finite measure obtained from an s-finite measure `μ`, such that
+`μ = μ.toFinite.withDensity μ.densityToFinite` (see `withDensity_densitytoFinite`). -/
+noncomputable def Measure.toFinite (μ : Measure α) [SFinite μ] : Measure α :=
+  Measure.sum (fun n ↦ (2 ^ (n + 1) * sFiniteSeq μ n Set.univ)⁻¹ • sFiniteSeq μ n)
```

Comment:
I think it'd be nice if you can filter out all the zero measures so that Measure.toFinite is a probability measure.

Reply:
But there might be only finitely many non-zero measures, so the sum of the `2 ^ (-(n+1))` would not be 1 anyway. Perhaps a better way to get that result is to simply rescale the resulting finite measure to turn in into a probability measure (unless it is zero of course, which happens if µ is zero).

Reply:
Done! `toFinite` is now a probability measure, unless it is zero.